File: 912.sort-an-array.go
/*
 * @lc app=leetcode.cn id=912 lang=golang
 * @lcpr version=
 *
 * [912] 排序数组
 */

// @lc code=start
func sortArray(nums []int) []int {
	quickSort(nums, 0, len(nums)-1)
	return nums
}

func partition(nums []int, left, right int) int {
	pivot := threeOfMid(nums, left, right)
	i, j := left, right
	for i < j {
		for i < j && nums[j] >= pivot {
			j--
		}
		for i < j && nums[i] <= pivot {
			i++
		}
		nums[i], nums[j] = nums[j], nums[i]
	}
	nums[left], nums[i] = nums[i], nums[left]
	return i
}

func threeOfMid(nums []int, left, right int) int {
	mid := left + (right-left)/2
	if nums[left] > nums[right] {
		nums[left], nums[right] = nums[right], nums[left]
	}
	if nums[mid] > nums[right] {
		nums[mid], nums[right] = nums[right], nums[mid]
	}
	if nums[mid] > nums[left] {
		nums[mid], nums[left] = nums[left], nums[mid]
	}
	return nums[left]
}

func quickSort(nums []int, left, right int) {
	if left >= right {
		return
	}

	pivot := partition(nums, left, right)
	quickSort(nums, left, pivot-1)
	quickSort(nums, pivot+1, right)
}

// @lc code=end

/*
// @lcpr case=start
// [5,2,3,1]\n
// @lcpr case=end

// @lcpr case=start
// [5,1,1,2,0,0]\n
// @lcpr case=end

// @lcpr case=start
// [-4,0,7,4,9,-5,-1,0,-7,-1]\n
// @lcpr case=end
*/

File: 912.sort-an-array.js
/*
 * @lc app=leetcode.cn id=912 lang=javascript
 * @lcpr version=
 *
 * [912] 排序数组
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortArray = function (nums) {
  quickSort(nums, 0, nums.length - 1);
  return nums;
};

function quickSort(nums, left, right) {
  if (left >= right) return;

  let pivotIdx = Math.floor(Math.random() * (right - left + 1)) + left;
  let pivot = nums[pivotIdx];

  let i = left;
  let j = right;
  while (i <= j) {
    while (nums[i] < pivot) {
      i++;
    }
    while (nums[j] > pivot) {
      j--;
    }

    if (i <= j) {
      [nums[i], nums[j]] = [nums[j], nums[i]];
      i++;
      j--;
    }
  }

  quickSort(nums, left, j);
  quickSort(nums, i, right);
}

function mergeSort(nums, left, right) {
  if (left >= right) return;

  let mid = Math.floor((left + right) / 2);
  mergeSort(nums, left, mid);
  mergeSort(nums, mid + 1, right);

  let i = left;
  let j = mid + 1;
  let temp = [];
  while (i <= mid && j <= right) {
    if (nums[i] <= nums[j]) {
      temp.push(nums[i]);
      i++;
    } else {
      temp.push(nums[j]);
      j++;
    }
  }

  while (i <= mid) {
    temp.push(nums[i]);
    i++;
  }

  while (j <= right) {
    temp.push(nums[j]);
    j++;
  }

  for (let i = left; i <= right; i++) {
    nums[i] = temp[i - left];
  }
}

// @lc code=end

/*
// @lcpr case=start
// [5,2,3,1]\n
// @lcpr case=end

// @lcpr case=start
// [5,1,1,2,0,0]\n
// @lcpr case=end

 */
